README.md
# MinorWork Framework

[![Build Status](https://travis-ci.org/CQD/minorwork.svg?branch=master)](https://travis-ci.org/CQD/minorwork)

A minimal PHP framework that is serious on being minimal.

Features:

- A container.
- Named routing with redirection support
- Less than 300 lines.
- Only external dependency is `nikic/fast-route` for routing.
- Optional Twig integration ([minorwork-twig](https://github.com/CQD/minorwork-twig))

## Install

You can install MinorWork with composer.

```
composer require minorwork/minorwork
```

PHP 5.4 or newer required.

## Usage

Using MinorWork framework looks like this:

```php
<?php
// This is index.php
// All traffic should bed handled by this file.
// You may need edit your .htaccess for this.

include __DIR__ . '/../vendor/autoload.php';

use MinorWork\App;

$app = new App();
$app->setRouting([
    'root'         => ['/', '\Controller\Index:home'],
    'loginForm'    => ['GET', '/login', '\Controller\Login:form'],
    'loginAction'  => ['POST', '/login', '\Controller\Login:login'],
    'article'      => ['/articles/{id}', '\Controller\Article:show'],
    'users'        => ['\Controller\User:list'], // matches /users
    'userProfile'  => ['/user/{id}', '\Controller\User:profile'],
]);

$app->run();
```

Define routing, the run the app. MinorWork will take care the rest.

Another example is in [example/](example/) folder. You can see it in action using PHP dev server.

```shell
cd example
php -S localhost:8765
```

Read the [documentation](doc/en/overview.md) for more detail.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented here.

## [0.4.3] - 2019006007
- Add `$app->next()` as alias of `$app->prependHandlerQueue()`

## [0.4.2] - 2019-05-01

### Added
- Add `$app->env` helper for easy access to environment variables.

## [0.4.1] - 2017-10-14

### Added
- Add `staticPath()` and `staticFullPath()`, to get path of static asset when project put into subfolder.

## [0.4.0] - 2017-09-10

### Changed
- Routing now works when project put into subfolder of HTTP doc root.


tests/bootstrap.php
<?php
namespace MinorWork;

include __DIR__ . '/../vendor/autoload.php';

class HeaderMock
{
    private static $headers = [];
    public static function add($header)
    {
        $key = strtoupper(strtok(':', $header));
        static::$headers[$key] = $header;
    }
    public static function remove($name = null)
    {
        if (null === $name) {
            static::$headers = [];
        } else {
            $key = strtoupper(strtok(':', $header));
            unset(static::$headers[$key]);
        }
    }
    public static function listAll()
    {
        return array_values(static::$headers);
    }
}

function header($header)
{
    HeaderMock::add($header);
}

function header_remove($name = null)
{
     HeaderMock::remove($name);
}


tests/src/AppTest.php
<?php
namespace MinorWork;

use PHPUnit\Framework\TestCase;
use PHPUnit\Runner\Version as PHPUnitVersion;

class AppTest extends TestCase
{
    public function setUp()
    {
        $_SERVER['SCRIPT_NAME'] = '/index.php';
        $_SERVER['HTTP_HOST'] = 'example.com';
        $_SERVER['HTTPS'] = 'off';
    }

    public function testCreationApp()
    {
        $app = new App;
        $this->assertInstanceOf('\MinorWork\Session\NativeSession', $app->session);
        $this->assertInstanceOf('\MinorWork\View\SimpleView', $app->view);
        $this->assertNull($app->notExistingField);

        $app = new App([
            'view' => '\MinorWork\View\JsonView',
            'newField' => 'Some Text',
        ]);
        $this->assertInstanceOf('\MinorWork\Session\NativeSession', $app->session);
        $this->assertInstanceOf('\MinorWork\View\JsonView', $app->view);
        $this->assertEquals('Some Text', $app->newField);
        $this->assertNull($app->notExistingField);
    }

    public function testSet()
    {
        $app = new App;

        // test factory function
        $invokeCnt = 0;
        $factory = function() use (&$invokeCnt) {
            $invokeCnt++;
            return ['A' => 4, 'B' => 2];
        };

        $app->a = $factory;
        $this->assertEquals(['A' => 4, 'B' => 2], $app->a);
        $this->assertEquals(['A' => 4, 'B' => 2], $app->a);

        $this->assertEquals(1, $invokeCnt, 'Factory function should only be called once');

        // test class name
        $app->b = '\stdClass';
        $this->assertInstanceOf('\stdClass', $app->b);

        // test simple set
        $app->c = 'MIEWMIEWMIEW';
        $this->assertEquals('MIEWMIEWMIEW', $app->c);

        // test override item
        $app->c = 'HAHAHA';
        $this->assertEquals('HAHAHA', $app->c);
    }

    public function testHandlerAlias()
    {
        $aCalled = false;

        $app = new App();
        $app->handlerAlias(['a' => function () use (&$aCalled) { $aCalled = true;},]);
        $app->setRouting(['routea' => ['a']]);
        $app->runAs('routea');

        $this->assertTrue($aCalled);
    }

    /**
     * @dataProvider routeProvider
     */
    public function testRoute($routes, $name, $expectedParams, $path)
    {
        if (null === $path) {
            return;
        }

        $app = new App();
        $app->setRouting($routes);

        $routeInfo = $app->route('GET', $path);
        if (null === $name) {
            $this->assertNull($routeInfo);
        } else {
            list($parsedName, $parsedParams) = $routeInfo;
            $expectedHandler = end($routes[$name]);
            $parsedHandler = end($routes[$parsedName]);
            $this->assertEquals($parsedName, $name);
            $this->assertEquals($parsedParams, $expectedParams);
            $this->assertEquals($parsedHandler, $expectedHandler);
        }
    }

    public function routeProvider()
    {
        $routes = $this->routes();
        return [
            [$routes, 'a', ['paramB' => 'YX', 'paramA' => 'XY'], '/a/XY/YX'],
            [$routes, 'b', ['paramB' => 'YX', 'paramA' => 'XY', 'paramC' => 'C'], '/b/XY/YX/C'],
            [$routes, 'b', ['paramA' => 'XY'], '/b/XY'],
            [$routes, 'd', [], '/d'],
            [$routes, 'slashhell', ['a' => 'A', 'b' => '42'], '///////42///A'],
            [$routes, null, ['paramB' => 'XY'], '/not/exists'],
        ];
    }

    public function routeProviderWithQueryString()
    {
        $routes = $this->routeProvider();
        foreach ($routes as &$route) {
            $query = ['query' => 'string', 'route' => $route[1] ?: 'null'];
            $route[3] .= '?' . http_build_query($query);
            $route[] = $query;
        }
        return $routes;
    }

    /**
     */
    public function testMultipleHandler()
    {
        $app = new App();
        $app->setRouting($this->routes());

        $params = ['a' => 42, 'B' => 'John'];

        ob_start();
        $app->runAs('multihandler', $params);
        $output = ob_get_clean();

        $this->assertTrue($app->handler1, "1st handler should execute.");
        $this->assertEquals($params, $app->handlerParams1);
        $this->assertEquals(null, $app->handlerPrevOutput1);

        $this->assertTrue($app->handler2, "2nd handler should execute.");
        $this->assertEquals($params, $app->handlerParams2);
        $this->assertEquals(['a' => 4], $app->handlerPrevOutput2);

        $this->assertTrue($app->handler3, "3rd handler should execute.");
        $this->assertEquals($params, $app->handlerParams3);
        $this->assertEquals(['a' => 4, 'b' => 2], $app->handlerPrevOutput3);

        $this->assertEquals('{"a":4,"b":2}', $output, "Should render json");
    }

    public function testRunAs()
    {
        $aCalled = false;

        $app = new App();
        $app->setRouting(['a' => [function() use (&$aCalled) {$aCalled = true;}]]);

        $app->runAs('a');
        $this->assertTrue($aCalled);

        if (method_exists($this, 'expectException')) {
            $this->expectException('\Exception');
            $app->runAs('non_exists_route');
        }
    }

    /**
     * @dataProvider routeProvider
     * @dataProvider routeProviderWithQueryString
     */
    public function testRun($routes, $name, $expectedParams, $path)
    {
        $app = new App();
        $app->setRouting($routes);

        ob_start();
        $parsedParams = $app->run([
            'method' => 'GET',
            'uri' => $path,
        ]);
        ob_end_clean();

        if (null === $name) {
            $this->assertNull($parsedParams);
            $this->assertEquals(404, http_response_code(), "No match found, default handler should send 404 status code");
        } else {
            $this->assertEquals($parsedParams, $expectedParams);
        }
    }

    /**
     * @dataProvider routeProvider
     * @dataProvider routeProviderWithQueryString
     * @dataProvider badRoutePathProvider
     */
    public function testRoutePath($routes, $name, $params, $expectedPath, $query = [])
    {
        $app = new App();
        $app->setRouting($routes);

        if (!method_exists($this, 'expectException') && (null === $name || null === $expectedPath)) {
            $this->markTestSkipped('expectException() not available.');
            return;
        }

        if (null === $name) {
            $name = 'not_exist_route';
            $this->expectException('\Exception', "Throw exception when route does not exists.");
        }

        if (null === $expectedPath) {
            $this->expectException('\Exception', "Throw exception when did not provide enough params to creath actual path");
        }

        $actualPath = $app->routePath($name, $params, $query);
        $this->assertEquals($expectedPath, $actualPath);

        $httpFullPath = "http://example.com{$actualPath}";
        $httpsFullPath = "https://example.com{$actualPath}";

        $_SERVER['HTTPS'] = false;
        $this->assertEquals($httpFullPath, $app->routeFullPath($name, $params, $query));

        $_SERVER['HTTPS'] = 'off';
        $this->assertEquals($httpFullPath, $app->routeFullPath($name, $params, $query));

        $_SERVER['HTTPS'] = 'on';
        $this->assertEquals($httpsFullPath, $app->routeFullPath($name, $params, $query));
    }

    public function badRoutePathProvider()
    {
        $routes = $this->routes();
        return [
            [$routes, 'a', [], null, []],
        ];
    }

    public function testStaticPath()
    {
        $app = new App();
        $this->assertEquals('/a.js', $app->staticPath('/a.js'));
        $this->assertEquals('http://example.com/a.js', $app->staticFullPath('/a.js'));

        $_SERVER['SCRIPT_NAME'] = '/subfolder/index.php';
        $app = new App();
        $this->assertEquals('/subfolder/a.js', $app->staticPath('/a.js'));
        $this->assertEquals('http://example.com/subfolder/a.js', $app->staticFullPath('/a.js'));
    }

    /**
     * @dataProvider routeProvider
     * @dataProvider routeProviderWithQueryString
     * @dataProvider badRedirectToProvider
     */
    public function testRedirectTo($routes, $name, $params, $expectedPath, $query = [])
    {
        if (!method_exists($this, 'expectException') && (null === $name || null === $expectedPath)) {
            $this->markTestSkipped('expectException() not available.');
            return;
        }

        if (null === $name) {
            $name = 'not_exist_route';
            $this->expectException('\Exception', "Throw exception when route does not exists.");
        }

        if (null === $expectedPath) {
            $this->expectException('\Exception', "Throw exception when did not provide enough params to creath actual path");
        }

        $app = new App();
        $app->setRouting($routes);

        $app->redirectTo($name, $params, $query);
        $this->assertContains("Location: {$expectedPath}", HeaderMock::listAll());
        header_remove();
    }

    public function badRedirectToProvider()
    {
        $routes = $this->routes();
        return [
            [$routes, 'a', [], null, []],
        ];
    }

    /**
     * @dataProvider executeHandlerProvider
     */
    public function testExecuteHandler($alias, $handler, $excpectedOutput, $shouldThrowException)
    {
        $app = new App();
        $app->handlerAlias($alias);


        if (!method_exists($this, 'expectException') && $shouldThrowException) {
            $this->markTestSkipped('expectException() not available.');
            return;
        }

        if ($shouldThrowException) {
            $this->expectException('\Exception');
        }

        $output = $app->executeHandler($handler, []);
        $this->assertEquals($excpectedOutput, $output);
    }

    public function executeHandlerProvider()
    {
        $alias = ['a' => '\MinorWork\MockController:sw'];
        return [
            // [alias list, handler, return value, throws exception]
            [$alias, function(){return 'Wonderful';}, 'Wonderful', false],
            [$alias, '\MinorWork\MockController:sw', 'Star Wars', false],
            [$alias, '\MinorWork\MockController:st', 'Star Trek', false],
            [$alias, 'a', 'Star Wars', null],
            [$alias, 'b', null, true],
            [$alias, new \stdClass, null, true],
        ];
    }

    public function testHandlerClassPrefix()
    {
        $app = new App();
        $app->handlerClassPrefix = "\MinorWork\\";
        $app->setRouting([
            'sw' => ['MockController:sw'],
            'st' => ['MockController:st'],
        ]);

        $this->assertEquals('Star Wars', $app->runAs('sw'));
        $this->assertEquals('Star Trek', $app->runAs('st'));
    }

    public function testDefaultErrorHandler()
    {
        $app = new App();
        $app->setRouting([
            'iWillFail' => [function($app, $params){throw new \Exception("Ha");}],
        ]);

        $options = [
            'method' => 'GET',
            'uri' => '/iWillFail',
        ];
        ob_start();
        $app->run($options);
        $output = ob_get_contents();
        ob_end_clean();

        $this->assertEquals('Oops, some thing is wrong!', $output);
        $this->assertEquals(500, http_response_code());
    }

    public function testHandlerQueue()
    {
        $app = new App();
        $app->handlerAlias('A', function(){echo 'a';});
        $app->handlerAlias('B', function(){echo 'b';});
        $app->handlerAlias('C', function(){echo 'c';});
        $app->handlerAlias('D', function(){echo 'd';});
        $app->handlerAlias('PA', function($app){$app->prependHandlerQueue('A');});
        $app->handlerAlias('PB', function($app){$app->prependHandlerQueue('B');});
        $app->handlerAlias('PC', function($app){$app->next('C');});
        $app->handlerAlias('PD', function($app){$app->next('D');});
        $app->handlerAlias('AA', function($app){$app->appendHandlerQueue('A');});
        $app->handlerAlias('AB', function($app){$app->appendHandlerQueue('B');});
        $app->handlerAlias('AC', function($app){$app->appendHandlerQueue('C');});
        $app->handlerAlias('AD', function($app){$app->appendHandlerQueue('D');});

        $custom = function($direction, $names) {
            return function($app) use ($direction, $names){
                foreach ($names as $name) {
                    $method = ('P' === $direction) ?  'prependHandlerQueue' : 'appendHandlerQueue';
                    $app->$method($name);
                }
            };
        };

        $stop = function($names = []) {
            return function($app) use ($names){
                $app->stop();
                foreach ($names as $name) {
                    $app->appendHandlerQueue($name);
                }
            };
        };

        $tests = [
            [['A', 'B', 'C', 'D'], 'abcd'],
            [['A', 'PB', 'C', 'D'], 'abcd'],
            [['A', 'AB', 'AC', 'D'], 'adbc'],
            [['A', $custom('A', ['D', 'A', 'C', 'B']), 'AC', 'D'], 'addacbc'],
            [['A', $custom('P', ['D', 'A', 'C', 'B']), 'AC', 'D'], 'abcaddc'],
            [['A', 'PB', 'AC', $stop(), 'A', 'D'], 'ab'],
            [['A', 'PB', 'AC', $stop(['B', 'C']), 'A', 'D'], 'abbc'],
        ];

        foreach ($tests as $idx => $test) {
            $app->setRouting([
                't' => [$test[0]],
            ]);
            ob_start();
            $app->runAs('t');
            $output = ob_get_contents();
            ob_end_clean();

            $no = $idx + 1;
            $this->assertEquals($test[1], $output, "Handler queue test {$no} should output `{$test[1]}`");
        }
    }

    public function testSubFolder()
    {
        $_SERVER['SCRIPT_NAME'] = '/subfolder/index.php';

        $routing = [
            'p1'      => ['/p1', 'p1:p1'],
            'default' => ['*',   'de:de'],
        ];

        $app = new App();
        $app->setRouting($routing);

        $this->assertEquals(['p1', []], $app->route('GET', '/subfolder/p1'));
        $this->assertEquals(null, $app->route('GET', '/subfolder/p11'));
        $this->assertEquals('/subfolder/p1', $app->routePath('p1'));
        $this->assertEquals('http://example.com/subfolder/p1', $app->routeFullPath('p1'));
    }

    public function testMethod()
    {
        $app = new App();
        $app->setRouting($this->routes());

        $this->assertEquals(['put', ['paramA' => 'p1']], $app->route('PUT', '/put/p1'));
        $this->assertEquals(['putget', ['paramA' => 'p2']], $app->route('PUT', '/putget/p2'));
        $this->assertEquals(['putget', ['paramA' => 'p3']], $app->route('GET', '/putget/p3'));
    }

    public function testEnv()
    {
        $app = new App();
        $this->assertNull($app->env->myenv);

        putenv('myenv=myvalue1');
        define('MYCONSTANT', 'myvalue2');
        $_ENV['myenv3'] = 'myvalue3';
        $this->assertSame('myvalue1', $app->env->myenv);
        $this->assertSame('myvalue2', $app->env->MYCONSTANT);
        $this->assertSame('myvalue3', $app->env->myenv3);

        putenv('myenv');
        unset($_ENV['myenv3']);
    }

    private function routes()
    {
        $echo = function($a, $p){return $p;};
        return [
            'a' => ['/a/{paramA}/{paramB}', $echo],
            'b' => ['/b/{paramA}[/{paramB}[/{paramC}]]', $echo],
            'c' => ['/basic/path', $echo],
            'd' => [$echo],
            'slashhell' => ['///////{b:\d+}///{a}', $echo],
            'multihandler' => [[
                function($a, $p, $po) {
                    $a->handler1 = true;
                    $a->handlerParams1 = $p;
                    $a->handlerPrevOutput1 = $po;
                    return ['a'=>4];
                },
                function($a, $p, $po) {
                    $a->handler2 = true;
                    $a->handlerParams2 = $p;
                    $a->handlerPrevOutput2 = $po;
                    return $po + ['b' => 2];
                },
                function($a, $p, $po) {
                    $a->handler3 = true;
                    $a->handlerParams3 = $p;
                    $a->handlerPrevOutput3 = $po;
                    $a->view->prepare(json_encode($po));
                    $a->stop();
                },
                function($a, $p){
                    throw new \Exception("Already stopped, third handler should not be called!");
                },
            ]],
            'put' => ['PUT', '/put/{paramA}', $echo],
            'putget' => [['PUT', 'GET'], '/putget/{paramA}', $echo],
        ];
    }
}

class MockController
{
    public function st(){ return "Star Trek";}
    public function sw(){ return "Star Wars";}
}


tests/src/View/JsonViewTest.php
<?php
namespace MinorWork\View;

use PHPUnit\Framework\TestCase;

class JsonViewTest extends TestCase
{
    public function testRender()
    {
        $v = new JsonView;

        $v->prepare(['a' => 'b']);
        $this->assertEquals((string) $v, '{"a":"b"}');

        $v->prepare(['a' => ['b' => 'c']]);
        $this->assertEquals((string) $v, '{"a":{"b":"c"}}');
    }
}


example/index.php
<?php
include __DIR__ . '/../vendor/autoload.php';

use MinorWork\App;

$app = new App();
$app->handlerAlias([
    'lookFor' => 'ExampleController:lookingFor',
]);
$app->setRouting([
    'root'       => ['/', 'itWorks'],
    'basic'      => ['itWorks'],
    'params'     => ['/p/{b1:\d+}[/{b2}]', 'itWorks'],
    'controller' => ['/c/{action}', ['itWorks', 'ExampleController:lookingFor']],
    'aliased'    => ['/a/{action}', ['itWorks', 'lookFor']],
    'simpleview' => ['simpleView'],
    'redirect'   => ['/r/{name}', function($app, $params){
        $app->redirectTo($params['name'], $app->get('_GET'));
    }],
    // 'default' => [function($app, $params){echo "Override default handler!";}],
]);

$app->run();

exit;

/////////////////////////////////////////

function simpleView($app, $params)
{
    $app->view->prepare(
        'I am a {job}, not a {not_my_job}.',
        ['job' => 'Doctor', 'not_my_job' => 'mechanic']
    );
}

function itWorks($app, $params)
{
    $template = <<<TEMPLATE
<pre><hr>[Links]

<a href='/basic'>/basic</a> The basic.
<a href='/p/12345/ParamForB'>/p/12345/ParamForB</a> URL parameters.
<a href='/not_exist'>/not_exist</a> Fallback to default handler when no match found.
<a href='/c/love'>/c/love</a> Controller class example.
<a href='/a/magic'>/a/magic</a> Aliased request handler
<a href='/simpleview'>/simpleview</a> SimpleView
<a href='/r/controller?action=peace'>/r/controller?action=peace</a> Supports named redirection.

<hr>[params]

TEMPLATE;
    $template .= json_encode($params, JSON_PRETTY_PRINT);
    $app->view->prepare($template);
}

class ExampleController
{
    public function lookingFor($app, $params)
    {
        $view = $app->view;
        $view->prepare("{$view}<hr>You are looking for {$params['action']}");
    }
}


src/View/SimpleView.php
<?php

namespace MinorWork\View;

class SimpleView
{
    private $template = '';
    private $params = [];

    /**
     * Prepare to render result
     *
     * The framework will call `toString()` to get rendered result.
     *
     * @param string $template
     * @param string $params
     */
    public function prepare($template, $params = [])
    {
        $this->template = $template;
        $this->params = $params;
    }

    /**
     * Actually render the template
     * @return string rendered result
     */
    public function __toString()
    {
        $template = $this->template;

        foreach ($this->params as $key => $value) {
            if (!is_scalar($value) && !method_exists($value, '__toString') ) {
                $value = '[' . gettype($value) . ']';
            }
            $keys[] = "{{$key}}";
            $values[] = $value;
        }

        return isset($keys) ? str_replace($keys, $values, $template) : $template;
    }
}


src/Session/NativeSession.php
<?php

namespace MinorWork\Session;

/**
 * PHP Native Session handler.
 */
class NativeSession implements SessionInterface
{
    const FLASHROOTKEY = 'MinorWorkInternal.flash';

    private $newFlash = [];
    private $oldFlash;

    public function __construct()
    {
        if (!isset($_SESSION)) {
            session_start();
        }

        if (!isset($_SESSION) && 'cli' === PHP_SAPI) {
            $_SESSION = []; // @codeCoverageIgnore
        }

        if (!isset($_SESSION)) {
            throw new \Exception("Could not init session!"); // @codeCoverageIgnore
        }

        $this->oldFlash = isset($_SESSION[self::FLASHROOTKEY]) ? $_SESSION[self::FLASHROOTKEY] : null;
    }

    public function __destruct()
    {
        if ($this->newFlash) {
            $_SESSION[self::FLASHROOTKEY] = $this->newFlash;
        } else {
            unset($_SESSION[self::FLASHROOTKEY]);
        }
    }

    public function get($key, $default = null)
    {
        // Check keys in the order of:
        // - normal key
        // - previous flash data
        // - current flash data
        foreach ([$_SESSION, $this->oldFlash, $this->newFlash] as $S) {
            if (isset($S[$key])) {
                return $S[$key];
            }
        }

        // No data found, return default value
        return $default;
    }

    public function getMany($keys)
    {
        $result = [];
        foreach ($keys as $key) {
            $result[$key] = $this->get($key);
        }
        return $result;
    }

    public function set($key, $value)
    {
        $this->setManyTo($_SESSION, [$key => $value]);
    }

    public function setMany($data)
    {
        $this->setManyTo($_SESSION, $data);
    }

    public function flash($key, $value)
    {
        $this->setManyTo($this->newFlash, [$key => $value]);
    }

    public function flashMany($data)
    {
        $this->setManyTo($this->newFlash, $data);
    }

    private function setManyTo(&$target, $data)
    {
        // Remove null data
        foreach ($data as $key => $value) {
            if (null === $value) {
                unset($target[$key]);
                unset($data[$key]);
            }
        }

        // set normal data in one shoot
        $target = array_replace($target, $data);
        return $target;
    }
}


src/App.php
<?php

namespace MinorWork;

use FastRoute\RouteParser\Std as RouteParserStd;

/**
 * Entry point for an application
 */
class App
{
    private $itemSetting = [];
    private $items = [];

    private $routings = null;
    private $router = null;
    private $dispatcher = null;

    private $handlerAlias = [];
    private $handlerQueue = [];

    public function __construct($items = [])
    {
        // Default container item
        $this->set($items + [
            '_GET' => $_GET,
            '_POST' => $_POST,
            '_SERVER' => $_SERVER,
            'view' => '\MinorWork\View\SimpleView',
            'session' => '\MinorWork\Session\NativeSession',
            'env' => '\MinorWork\Helper\Env',
        ]);

        // check subfolder
        $scriptName = @$this->get('_SERVER')['SCRIPT_NAME'];
        $this->items['___.baseDir'] = dirname($scriptName) ?: '/';

        // Default routings
        $this->routings = [
            'default' => ['*', '*', function($app, $params){
                http_response_code(404);
                $app->set('view', '404! I can not find what you want!');
            }],
            'defaultError' => ['!', '!', function ($app, $params) {
                syslog(LOG_ERR, end($params));
                http_response_code(500);
                $app->set('view', 'Oops, some thing is wrong!');
            }],
        ];
    }

    /**
     * Get an item from container
     * @param string $name name for desired item
     */
    public function get($name)
    {
        if (isset($this->items[$name])) {
            return $this->items[$name];
        }

        if (!isset($this->itemSetting[$name])) {
            return null;
        }

        $itemSetting = $this->itemSetting[$name];
        if (is_callable($itemSetting)) {
            $item = $itemSetting();
        } elseif (is_string($itemSetting) && class_exists($itemSetting)) {
            $item = new $itemSetting;
        } else {
            $item = $itemSetting;
        }

        $this->items[$name] = $item;
        return $item;
    }

    public function __get($name)
    {
        return $this->get($name);
    }

    /**
     * Set one or multiple items, or factory function of items, or class of items, into DI container.
     */
    public function set($name, $value = null)
    {
        $values = is_array($name) ? $name : [$name => $value];
        foreach ($values as $name => $value) {
            $this->itemSetting[$name] = $value;
        }

        unset($this->items[$name]);
    }

    public function __set($name, $value)
    {
        return $this->set($name, $value);
    }

    /**
     * Set app routing.
     *
     * TODO group routing support
     */
    public function setRouting(array $routings = [])
    {
        $this->dispatcher = null;
        $this->routings = $routings + $this->routings;

        foreach ($this->routings as $name => &$routing) {
            if (1 == count($routing)) {
                array_unshift($routing, '/' . ltrim($name, '/'));
            }
            if (2 == count($routing)) {
                array_unshift($routing, ['GET', 'POST']);
            }
        }
    }

    /**
     * Set alias name for request handlers
     */
    public function handlerAlias($name, $value = null)
    {
        $alias = is_array($name) ? $name : [$name => $value];
        $this->handlerAlias = $alias + $this->handlerAlias;
    }

    /**
     * Stop excuting next request handler
     */
    public function stop()
    {
        $this->handlerQueue = [];
    }

    /**
     * Determin which route to use and parameters in uri.
     *
     * @param string $method HTTP Method used
     * @param string $uri request path, without query string
     * @return array|null [$routeName, $params]
     */
    public function route($method, $uri)
    {
        // Remove subfolder from URI if needed
        $baseDir = $this->get('___.baseDir');
        if ("/" !== $baseDir && 0 === strpos($uri, $baseDir)) {
            $uri = substr($uri, strlen($baseDir));
        }

        // Check which route is being used.
        $routings = $this->routings;
        $this->dispatcher = $this->dispatcher ?: \FastRoute\simpleDispatcher(function(\FastRoute\RouteCollector $r) use ($routings) {
            foreach ($routings as $name => $routing) {
                // 'GET', '/user/{id:\d+}', $routeName
                $r->addRoute($routing[0], $routing[1], $name);
            }
        });

        $routeInfo = $this->dispatcher->dispatch($method, $uri);

        if (\FastRoute\Dispatcher::FOUND === $routeInfo[0]) {
            return array_slice($routeInfo, 1);
        }

        return null;
    }

    public function redirectTo($routeName, $params = [], $query = [])
    {
        header(sprintf("Location: %s", $this->routePath($routeName, $params, $query)));
        $this->stop();
    }

    public function routePath($routeName, $params = [], $query = [])
    {
        if (!isset($this->routings[$routeName])) {
            throw new \Exception("Route name '{$routeName}' not found!");
        }

        $pattern = $this->routings[$routeName][1];
        $path = $this->patternToPath($pattern, $params);
        if ($query) {
            $path .= "?" . http_build_query($query);
        }

        $path = rtrim($this->get('___.baseDir'), '/') . $path;

        return $path;
    }

    public function routeFullPath($routeName, $params = [], $query = [])
    {
        $isHttps = @$_SERVER['HTTPS'] && ('off' !== $_SERVER['HTTPS']);
        $schema = $isHttps ? "https://" : "http://";
        $host = @$_SERVER['HTTP_HOST'] ?: 'localhost';
        return $schema . $host . $this->routePath($routeName, $params, $query);
    }

    public function staticPath($path)
    {
        return rtrim($this->get('___.baseDir'), '/') . $path;
    }

    public function staticFullPath($path)
    {
        $isHttps = @$_SERVER['HTTPS'] && ('off' !== $_SERVER['HTTPS']);
        $schema = $isHttps ? "https://" : "http://";
        $host = @$_SERVER['HTTP_HOST'] ?: 'localhost';
        return $schema . $host . $this->staticPath($path);
    }
    /**
     * Convert url pattern to actual path
     */
    private function patternToPath($pattern, $params)
    {
        // Inspired by https://github.com/nikic/FastRoute/issues/66
        // but uses named params, don't care param order and size
        $routeParser = new RouteParserStd;
        $routes = $routeParser->parse($pattern);

        // check from the longest route to the shortest route
        // first (longest) full match is the route we need
        for ($i = count($routes) - 1; $i >= 0; $i--) {
            $url = '';
            foreach ($routes[$i] as $part) {
                // replace placeholder to actual value
                if (is_array($part)) {
                    $part = @$params[$part[0]];
                }

                // if route contains part not defined in $params, abandan this route
                if (null === $part) {
                    continue 2;
                }
                $url .= $part;
            }

            // first full match, this is our hero url
            return $url;
        }

        // no full match, throw Exception
        throw new \Exception(sprintf("Can't make path for '%s' with param: %s.", $pattern, json_encode($params)));
    }

    /**
     * Entry point for application
     */
    public function run($options = [])
    {
        $method = @$options['method'] ?: $_SERVER['REQUEST_METHOD'];

        $uri = @$options['uri'] ?: rawurldecode($_SERVER['REQUEST_URI']);
        if (false !== $pos = strpos($uri, '?')) {
            $uri = substr($uri, 0, $pos);
        }

        $routeInfo = $this->route($method, $uri);
        if (!$routeInfo) {
            $routeInfo = ['default', []];
        }

        list($routeName, $params) = $routeInfo;

        try {
            return $this->runAs($routeName, $params);
        } catch (\Exception $e) {
            $params[] = $e;
            return $this->runAs('defaultError', $params);
        }
    }

    /**
     * Handle current request using specificed route handler
     */
    public function runAs($routeName, $params = [])
    {
        if (!isset($this->routings[$routeName])) {
            throw new \Exception("Route name '{$routeName}' not found!");
        }

        // populate handler queue
        $handler = end($this->routings[$routeName]);
        $this->appendHandlerQueue($handler);

        // Run everything in handler queue
        $output = null;
        while ($handler = array_shift($this->handlerQueue)) {
            $output = $this->executeHandler($handler, $params, $output);
        }
        echo $this->get('view');

        return $output;
    }

    /**
     * pupolate handler queue from the end
     */
    public function appendHandlerQueue($handler)
    {
        $this->addToHandlerQueue($handler, 'end');
    }

    /**
     * pupolate handler queue from the end
     */
    public function prependHandlerQueue($handler)
    {
        $this->addToHandlerQueue($handler, 'head');
    }

    /**
     * pupolate handler queue from the end
     */
    public function next($handler)
    {
        $this->addToHandlerQueue($handler, 'head');
    }

    private function addToHandlerQueue($handler, $order)
    {
        $isArrayOfHandler = (!is_callable($handler) && is_array($handler));
        $handlers = $isArrayOfHandler ? $handler : [$handler];

        if ('end' === $order) {
            foreach ($handlers as $handler) {
                $this->handlerQueue[] = $handler;
            }
        } else {
            foreach (array_reverse($handlers) as $handler) {
                array_unshift($this->handlerQueue, $handler);
            }
        }
    }

    /**
     * Parse and execute request handler
     */
    public function executeHandler($handler, $params, $prevOutput = null)
    {
        // a function(-ish) thing which can be called
        if (is_callable($handler)) {
            return $handler($this, $params, $prevOutput);
        }

        // aliased handler
        if ($actualHandler = @$this->handlerAlias[$handler]) {
            return $this->executeHandler($actualHandler, $params, $prevOutput);
        }

        // only callable and controller/method pair (`:` seprated string) can be accepted
        if (!is_string($handler)) {
            $type = gettype($handler);
            throw new \Exception("'{$type}' can not be used as handler, should be a callable or controller/method pair!");
        }

        if (false === $pos = strpos($handler, ':')) {
            throw new \Exception("Can not parse '{$handler}' for corresponding controller/method!");
        }

        // controller/method pair
        list($class, $method) = explode(':', $handler, 2);
        $class = $this->handlerClassPrefix ? "{$this->handlerClassPrefix}{$class}" : $class;
        $controller = new $class;
        return $controller->$method($this, $params, $prevOutput);
    }
}
